sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,276 @@
+It was generated only on kernels earlier. There are two reasons to start
+generating this attribute on `SYCL_EXTERNAL` functions:
+
+- Later in pipeline, this attribute will be used by `sycl-post-link` tool to
+  separate `SYCL_EXTERNAL` functions from non-`SYCL_EXTERNAL` functions with
+  external linkage.
+- `module-id` attribute also contains information about source file where the
+  function comes from. This information will be used to perform device code
+  split on device images that contain only exported functions.
+
+### sycl-post-link changes
+
+`sycl-post-link` performs 3 important tasks:
+- Arranges `SYCL_EXTERNAL` functions into a separate device image(s)
+- Supplies device images containing exports with an information about exported
+  symbols
+- Supplies each device image with an information about imported symbols
+
+`sycl-post-link` outlines `SYCL_EXTERNAL` functions with all their reachable
+dependencies (functions with definitions called from `SYCL_EXTERNAL` ones)
+into a separate device image(s) in order to create minimal self-contained
+device images that can be linked from the user's app. There are several
+notable moments though.
+
+If a `SYCL_EXTERNAL` function is used within a kernel defined in a shared
+library, it will be duplicated: one instance will be stored in the kernel's
+device image and the function won't exported from this device image, while the
+other will be stored in a special device image for other `SYCL_EXTERNAL`
+functions and will be marked as exported there. Such duplication is need for
+two reasons:
+- We aim to make device images with kernels self-contained so no JIT linker
+  invocations would be needed if we have definitions of all called functions.
+  Also note that if AOT is requested, it would be impossible to link anything
+  at runtime.
+- We could export `SYCL_EXTERNAL` functions from device images with kernels,
+  but it would mean that when user's app calls `SYCL_EXTERNAL` function, it has
+  to link a whole kernel and all its dependencies - not only it increases the
+  amount of unnecessary linked code, but might also lead to build errors if the
+  kernel uses some features, which are not supported by target device (and they
+  are not used in the `SYCL_EXTERNAL` function).
+Besides separating `SYCL_EXTERNAL` functions from kernels, they can be further
+split into separate device images if device code split is requested. This is
+done by grouping them using `module-id` attribute. Non-`SYCL_EXTERNAL` functions
+used by `SYCL_EXTERNAL` functions with different `module-id` attributes are
+copied to device images corresponding to those `SYCL_EXTERNAL` functions
+to make them self-contained
+In case one `SYCL_EXTERNAL` function uses another `SYCL_EXTERNAL` function
+with different `module-id` attribute, the second one is not copied to the
+device image with the first function, but dependency between those device images
+is recorder instead.
+
```

Comment:
Do we leave too much logic at sycl-post-link discretion here? My suggestions would be:
- don't make distinction between `SYCL_EXTERNAL` functions and kernels as long as code splitting is concerned. They are both external linkage items in the binary image. E.g. this should not be the case and sycl-post-link should be fixed:
  > if `SYCL_EXTERNAL` function is not called by any kernel, it won't be included into any device images.

- separation of `SYCL_EXTERNAL` functions from kernels as well as grouping into different binary images should ideally be user's responsibility and under his/her control, like in usual C++ tool chain. Otherwise I think post-link does too much magic behind the scenes.

I think it would be also good to describe a typical user scenario in more details from start to end to understand efficiency aspects. E.g. a users creates a shared library with a bunch of `SYCL_EXTERNAL` functions (command example), which are used to resolve dynamically linked symbols for both host and device. User may request to do code-split - per-top-level-function or per TU. What happens at runtime - resolution, linkage,...

If a kernel uses external function `A`, but the dynamically linked SPIRV module contains `A` and `B` - then `B` will be linked in as well and unnecessarily JITted? Can we avoid that via e.g. requiring  llvm-link-like `-only-needed` functionality in the underlying linker, or first translating to LLVM IR, linking with `llvm-link -only-needed` then translating back to SPIRV?

Reply:
> separation of `SYCL_EXTERNAL` functions from kernels as well as grouping into different binary images should ideally be user's responsibility and under his/her control, like in usual C++ tool chain.

It is not clear to me how C++ users can choose which binary images their functions go to if there is only one binary image in regular C++ within a shared library? Could you please clarify what do you mean by `binary images` here in context of regular C++?

> If a kernel uses external function A, but the dynamically linked SPIRV module contains A and B - then B will be linked in as well and unnecessarily JITted? 

This is outside of our control: JIT linkage is done by underlying runtimes and not by DPC++ RT.

>  Can we avoid that via e.g. requiring llvm-link-like -only-needed functionality in the underlying linker, or first translating to LLVM IR, linking with llvm-link -only-needed then translating back to SPIRV?

I guess we can, but it should really be discussed somewhere else - this document describes DPC++ Toolchain design, not how low-level runtimes work

Reply:
> It is not clear to me how C++ users can choose which binary images their functions go to if there is only one binary image in regular C++ within a shared library? Could you please clarify what do you mean by binary images here in context of regular C++?

By binary image I mean device binary (SPIRV) with accompanying data (e.g. symbols). Couple of ways to manually split into binaries:
- splitting the code into different .so/.dll via command line - user may create as many as needed
- semi-manual assembly of device binaries from SPIRV and the accompanying data using clang-offload-wrapper and subsequent linkage.

> I guess we can, but it should really be discussed somewhere else - this document describes DPC++ Toolchain design, not how low-level runtimes work.

This is a question of the design efficiency, so I don't see why it should not be discussed here. Toolchain can help via invoking underlying tools in efficient manner.

I believe a lot questions will go away if the design doc provides some more info I requested above:
- user scenario (preferably with multiple libraries involved)
- tool chain and RT processing - what module IDs are created, import/export lists, runtime dependence resolution and linkage.
Can you folks please add these details?

Reply:
I vote for arranging of `SYCL_EXTERNAL` functions to separate device images. I think it is better for size of device code being compiled. What if some application has dependency on one library, device image that provides required `SYCL_EXTERNAL` definition contains not requested kernel with some dependency on other library? We may end up JITting literally everything.
It also seems simpler from cache point of view, since if `SYCL_EXTERNAL`-only modules do not contain kernels we won't accidentally compile some unnecessary kernels while resolving some dependency, so we don't need to care about them in cache.

Reply:
The design was updated so we don't arrange SYCL_EXTERNAL functions to a separate device image. Instead, sycl-post-link will consider them as entry points during device code split.